File: src/main/java/com/zoarial/iot/models/actions/IoTActionList.java
package com.zoarial.iot.models.actions;

import java.util.*;

public class IoTActionList extends ArrayList<IoTAction> {
    final private HashMap<UUID, IoTAction> uuidMap;

    public IoTActionList(int len) {
        super(len);
        uuidMap = new HashMap<>(len);
    }

    public IoTActionList() {
        super();
        uuidMap = new HashMap<>();
    }

    public IoTActionList(Collection<? extends IoTAction> c) {
        super(c);
        uuidMap = new HashMap<>();
        for(IoTAction action : c) {
            uuidMap.put(action.getUuid(), action);
        }
    }

    @Override
    public boolean add(IoTAction action) {
        boolean b = super.add(action);
        if(b) {
            uuidMap.put(action.getUuid(), action);
        }
        return b;
    }

    public boolean contains(UUID uuid)  {
        return uuidMap.containsKey(uuid);
    }

    public IoTAction get(UUID uuid) {
        return uuidMap.get(uuid);
    }
}

File: src/main/java/com/zoarial/iot/action/model/IoTActionList.kt
package com.zoarial.iot.action.model

import java.util.*

class IoTActionList : ArrayList<IoTAction> {
    private val uuidMap: HashMap<UUID, IoTAction>

    constructor(len: Int) : super(len) {
        uuidMap = HashMap(len)
    }

    constructor() : super() {
        uuidMap = HashMap()
    }

    constructor(c: Collection<IoTAction>) : super(c) {
        uuidMap = HashMap<UUID, IoTAction>()
        for (action in c) {
            uuidMap[action.uuid!!] = action
        }
    }

    fun remove(uuid: UUID): IoTAction? {
        val index = indexOf(uuid)
        return if (index != -1) {
            uuidMap.remove(uuid)
            removeAt(index)
        } else {
            null
        }
    }

    fun indexOf(uuid: UUID): Int {
        val action = get(uuid)
        return action?.let { indexOf(it) } ?: -1
    }

    override fun add(action: IoTAction): Boolean {
        val b = super.add(action)
        if (b) {
            uuidMap[action.uuid!!] = action
        }
        return b
    }

    operator fun contains(uuid: UUID): Boolean {
        return uuidMap.containsKey(uuid)
    }

    operator fun get(uuid: UUID): IoTAction? {
        return uuidMap[uuid]
    }
}